### PR TITLE
Adding option to configure cgroup to start cri-containerd

### PR DESCRIPTION
**Pull request by abhinandanpb**
Manual test output:

```
ps -aux | grep cri-containerd
root     26047  7.0  0.7 224224 28872 pts/1    Sl+  15:11   0:00 ./cri-containerd --cgroup-path /containerd

$ cat /proc/26047/cgroup
11:net_cls,net_prio:/containerd
10:perf_event:/containerd
9:pids:/containerd
8:hugetlb:/containerd
7:memory:/containerd
6:freezer:/containerd
5:blkio:/containerd
4:devices:/containerd
3:cpuset:/containerd
2:cpu,cpuacct:/containerd
1:name=systemd:/containerd
```

Signed-off-by: Abhinandan Prativadi <abhi@docker.com>

### CODE REVIEW

File: pkg/server/service.go [hunk 1]
```diff
@@ -82,18 +85,37 @@ type criContainerdService struct {
 	client *containerd.Client
 	// streamServer is the streaming server serves container streaming request.
 	streamServer streaming.Server
+	// cgroup in which the cri-containerd is placed in
+	cgroup cgroups.Cgroup
```
Comment (Random-Liu) [inline pkg/server/service.go]: Seems like that we only need the path (as the backup for parent cgroup). Can we only save the path?

File: pkg/server/service.go [hunk 2]
```diff
@@ -82,18 +85,37 @@ type criContainerdService struct {
...
 		return nil, fmt.Errorf("failed to initialize containerd client with endpoint %q: %v", containerdEndpoint, err)
 	}
+	if cgroupPath != "" {
+		cg, err = cgroups.Load(cgroups.V1, cgroups.StaticPath(cgroupPath))
+		if err != nil {
+			if err != cgroups.ErrCgroupDeleted {
+				return nil, err
+			}
+			if cg, err = cgroups.New(cgroups.V1, cgroups.StaticPath(cgroupPath), &specs.LinuxResources{}); err != nil {
+				return nil, err
```
Comment (Random-Liu) [inline pkg/server/service.go]: ditto.

File: pkg/server/service.go [hunk 3]
```diff
@@ -82,18 +85,37 @@ type criContainerdService struct {
...
+			}
+		}
+		if err := cg.Add(cgroups.Process{
+			Pid: os.Getpid(),
```
Comment (Random-Liu) [inline pkg/server/service.go]: ditto.

File: pkg/server/service.go [hunk 4]
```diff
@@ -82,18 +85,37 @@ type criContainerdService struct {
...
+	streamAddress, streamPort, cgroupPath string) (CRIContainerdService, error) {
 	// TODO(random-liu): [P2] Recover from runtime state and checkpoint.
 
+	var cg cgroups.Cgroup
 	client, err := containerd.New(containerdEndpoint, containerd.WithDefaultNamespace(k8sContainerdNamespace))
 	if err != nil {
 		return nil, fmt.Errorf("failed to initialize containerd client with endpoint %q: %v", containerdEndpoint, err)
 	}
+	if cgroupPath != "" {
+		cg, err = cgroups.Load(cgroups.V1, cgroups.StaticPath(cgroupPath))
```
Comment (Random-Liu) [inline pkg/server/service.go]: nit: Can we move this logic into a separate function? Just keep the code cleaner.

File: vendor.conf [hunk 5]
```diff
@@ -3,6 +3,8 @@ github.com/boltdb/bolt v1.3.0-58-ge9cf4fa
 github.com/containerd/containerd cf09e32618398fc59fcb45bcfe9b4c0335972733
 github.com/containerd/continuity cf279e6ac893682272b4479d4c67fd3abf878b4e
 github.com/containerd/fifo fbfb6a11ec671efbe94ad1c12c2e98773f19e1e6
+github.com/containerd/cgroups 7a5fdd8330119dc70d850260db8f3594d89d6943
```
[the remark this answers is not in the document]
Comment (abhinandanpb) [reply vendor.conf]: cgroups doesn't vendor dependency. In such cases we will need to manually vendor the external dependencies. 
I think it should be ok. But I can  check it to match with containerd.

File: pkg/server/service.go (hunk 4 above)
Comment (abhinandanpb) [reply pkg/server/service.go]: will do so

File: pkg/server/service.go [hunk 6]
```diff
@@ -82,18 +85,37 @@ type criContainerdService struct {
...
+		if err != nil {
+			if err != cgroups.ErrCgroupDeleted {
+				return nil, err
```
[the remark this answers is not in the document]
Comment (abhinandanpb) [reply pkg/server/service.go]: will update

**Comment by Random-Liu** [comment]
LGTM overall with nits.

**Comment by abhinandanpb** [comment]
@Random-Liu addressed comments. PTAL.

File: pkg/server/helpers.go [hunk 7]
```diff
@@ -352,3 +354,23 @@ func resolveSymbolicLink(path string) (string, error) {
 	}
 	return filepath.EvalSymlinks(path)
 }
+
+// LoadCgroup loads the cgroup associated with path if it exits. If the cgroup
+// is not created it is created and returned.
+func LoadCgroup(cgroupPath string) (cgroups.Cgroup, error) {
```
Comment (Random-Liu) [inline pkg/server/helpers.go]: s/ LoadCgroup loads the cgroup associated with path if it exits/ loadCgroup loads the cgroup associated with path if it exits, and move the current process into the cgroup.

nit: Make it private if we don't want to expose it.

**Comment by Random-Liu** [comment]
LGTM. Will merge after test passes.

**Comment by Random-Liu** [comment]
LGTM. Will merge after test passes.

**Comment by abhinandanpb** [comment]
@Random-Liu Cool! Thanks for reviewing

**Pull request merged**